src/pages/configurePatch/configurePatchCore/ConfigureTasks/utils.ts

```diff
@@ -0,0 +1,192 @@
+import {
+  AliasState,
+  ChildPatchAliased,
+  CheckboxState,
+} from "pages/configurePatch/configurePatchCore/types";
+
+interface DeduplicateTasksResult {
+  [task: string]: CheckboxState;
+}
+
+type VariantTaskArray = {
+  [task: string]: boolean;
+}[];
+
+/**
+ * `deduplicateTasks` takes an array of objects containing the tasks for each build variant and the tasks from the previously configured patch
+ * @param currentBVTasks - an array of objects containing the tasks for each build variant
+ * @param previousActivatedBVTasks - an array of objects containing the tasks for each build variant from the previously configured patch
+ * @returns - an object containing the deduplicated tasks for each build variant and the state of each task
+ * - If a task has been previously configured and is present in the current build variant it is disabled
+ */
+const deduplicateTasks = (
+  currentBVTasks: VariantTaskArray,
+  previousActivatedBVTasks?: VariantTaskArray
+): DeduplicateTasksResult => {
+  const visibleTasks: DeduplicateTasksResult = {};
+  currentBVTasks.forEach((bv, bvIndex) => {
+    const previousBuildVariantTasks = previousActivatedBVTasks?.[bvIndex] || {};
+    Object.entries(bv).forEach(([taskName, value]) => {
+      switch (visibleTasks[taskName]) {
+        case CheckboxState.UNCHECKED:
+          // If a task is UNCHECKED and the next task of the same name is CHECKED it is INDETERMINATE
+          visibleTasks[taskName] = value
+            ? CheckboxState.INDETERMINATE
+            : CheckboxState.UNCHECKED;
+          break;
+        case CheckboxState.CHECKED:
+          // If a task is CHECKED and the next task of the same name is UNCHECKED it is INDETERMINATE
+          visibleTasks[taskName] = value
+            ? CheckboxState.CHECKED
+            : CheckboxState.INDETERMINATE;
+          break;
+        case (CheckboxState.INDETERMINATE,
+        CheckboxState.DISABLED_INDETERMINATE):
+          // If a task is INDETERMINATE or DISABLED_INDETERMINATE because of previous task statuses
+          // it wouldn't change when subsequent statuses are considered
+          break;
+        case CheckboxState.DISABLED_CHECKED:
+          // If a task is DISABLED_CHECKED and the next task of the same name is UNCHECKED it is DISABLED_INDETERMINATE
+          visibleTasks[taskName] = value
+            ? CheckboxState.DISABLED_CHECKED
+            : CheckboxState.DISABLED_INDETERMINATE;
+          break;
+
+        default:
+          visibleTasks[taskName] = value
+            ? CheckboxState.CHECKED
+            : CheckboxState.UNCHECKED;
+          break;
+      }
+      if (previousBuildVariantTasks[taskName]) {
+        visibleTasks[taskName] = isCheckboxIndeterminate(visibleTasks[taskName])
+          ? CheckboxState.DISABLED_INDETERMINATE
+          : CheckboxState.DISABLED_CHECKED;
```

Comment:
Combining the checkboxes' disabled and checked states kinda feels like a derived state antipattern to me. Was there any particular benefit to this? It seems like it would straightforward to do this check in the component itself and apply the disabled prop that way.

Reply:
Got it, that's fair enough! Another lightweight idea may be to separate the states? I feel like this would be pretty flexible. But definitely respect if you want to ticket this as later work instead!

```js
interface DeduplicateTasksResult {
  [task: string]: {
    checkbox: CheckboxState;
    previouslySeen: boolean;
}
```